community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java

```diff
@@ -0,0 +1,549 @@
+    public void close( boolean populationCompletedSuccessfully )
+    {
+        for ( int i = 0; i < populators.size(); )
+        {
+            Populator populator = populators.get( i );
+            try
+            {
+                populator.close( populationCompletedSuccessfully );
+                i++;
+            }
+            catch ( Throwable t )
+            {
+                fail( i, t );
+            }
+        }
+    }
+
+    @Override
+    public void markAsFailed( String failure ) throws IOException
+    {
+        throw new UnsupportedOperationException( "Multiple index populator can't be marked as failed." );
+    }
+
+    @Override
+    public long sampleResult( Out result )
+    {
+        throw new UnsupportedOperationException( "Multiple index populator can't perform index sampling." );
+    }
+
+    public void replaceIndexCounts( long uniqueElements, long maxUniqueElements, long indexSize )
+    {
+        populators.forEach( populator ->
+                storeView.replaceIndexCounts( populator.getDescriptor(), uniqueElements,
+                        maxUniqueElements, indexSize ) );
+    }
+
+    public void flipAfterPopulation()
+    {
+        for ( int i = 0; i < populators.size(); )
+        {
+            Populator populator = populators.get( i );
+            try
+            {
+                populator.flip();
+                i++;
+            }
+            catch ( Throwable t )
+            {
+                fail( i, t );
+            }
+        }
+    }
+
+    private int[] propertyKeyIds()
+    {
+        int[] ids = new int[populators.size()];
+        for ( int i = 0; i < populators.size(); i++ )
+        {
+            ids[i] = populators.get( i ).getPropertyKeyId();
+        }
+        return ids;
+    }
+
+    private int[] labelIds()
+    {
+        int[] ids = new int[populators.size()];
+        for ( int i = 0; i < populators.size(); i++ )
+        {
+            ids[i] = populators.get( i ).getLabelId();
+        }
+        return ids;
+    }
+
+    public void cancel()
+    {
+        replaceIndexCounts( 0, 0, 0 );
+        close( false );
+    }
+
+    private void populateFromQueueIfAvailable( final long currentlyIndexedNodeId )
+    {
+        if ( !queue.isEmpty() )
+        {
+            try ( MultipleIndexUpdater updater = newPopulatingUpdater( storeView ) )
+            {
+                do
+                {
+                    // no need to check for null as nobody else is emptying this queue
+                    NodePropertyUpdate update = queue.poll();
+                    // TODO: We see updates twice here from IndexStatisticsTest
+                    if ( update.getNodeId() <= currentlyIndexedNodeId )
+                    {
+                        updater.process( update );
+                    }
+                }
+                while ( !queue.isEmpty() );
+            }
+        }
+    }
+
+    private static class MultipleIndexUpdater implements IndexUpdater
```

Comment:
I don't think it makes sense to make them top-level since `MultiIndexUpdate` have intricate and private interaction with its parent, something that should not be visible from outside. Also `Populator` (now renamed to `IndexPopulation`) is mostly a grouping of some arguments for easier management of multiple such.